### PR TITLE
vsc.utils.run: improve handling tasks and pgids (HPC-5756)

### PR DESCRIPTION
**Pull request by stdweird**

### CODE REVIEW

File: lib/vsc/utils/run.py [hunk 1]
```diff
@@ -384,25 +384,51 @@ def _killtasks(self, tasks=None, sig=signal.SIGKILL, kill_pgid=False):
...
+                res = fn(*args)
             except OSError, err:
                 # ERSCH is no such process, so no issue
-                if not err.errno == errno.ESRCH:
-                    self.log.error("Failed to kill %s: %s" % (pid, err))
+                if err.errno != errno.ESRCH:
+                    self.log.error("Failed to %s from %s: %s" % (msg, pid, err))
             except Exception, err:
-                self.log.error("Failed to kill %s: %s" % (pid, err))
+                self.log.error("Failed to %s from %s: %s" % (msg, pid, err))
```
Comment (boegel) [inline lib/vsc/utils/run.py]: maybe something like:

```python
except Exception as err:
    if isinstance(err, OSError):
        ...
    else:
       ...
```

**Comment by boegel** [comment]
Looks OK, but no enhanced tests?

**Comment by JensTimmerman** [comment]
jenkins test this please

File: lib/vsc/utils/run.py [hunk 2]
```diff
@@ -384,25 +384,51 @@ def _killtasks(self, tasks=None, sig=signal.SIGKILL, kill_pgid=False):
...
+            res = None
+            try:
+                res = fn(*args)
             except OSError, err:
                 # ERSCH is no such process, so no issue
-                if not err.errno == errno.ESRCH:
-                    self.log.error("Failed to kill %s: %s" % (pid, err))
+                if err.errno != errno.ESRCH:
+                    self.log.error("Failed to %s from %s: %s" % (msg, pid, err))
             except Exception, err:
```
Comment (JensTimmerman) [inline lib/vsc/utils/run.py]: `Exception as err`

File: lib/vsc/utils/affinity.py [hunk 3]
```diff
@@ -153,7 +153,7 @@ def get_cpus(self):
         """
         self.cpus = []
         for bitmask in getattr(self, '__bits'):
-            for idx in xrange(NCPUBITS):
+            for _ in range(NCPUBITS):
```
Comment (JensTimmerman) [inline lib/vsc/utils/affinity.py]: why not xrange?

File: lib/vsc/utils/affinity.py [hunk 4]
```diff
@@ -238,7 +238,7 @@ def sched_getcpu():
 #    tobin not used anymore
 def tobin(s):
     """Convert integer to binary format"""
-    ## bin() missing in 2.4
+    # bin() missing in 2.4
```
Comment (JensTimmerman) [inline lib/vsc/utils/affinity.py]: we don't need to support python2.4 anymore, so actually this tobin can be dropped?

File: lib/vsc/utils/generaloption.py [hunk 5]
```diff
@@ -1705,7 +1706,7 @@ class SimpleOptionParser(ExtOptionParser):
 class SimpleOption(GeneralOption):
     PARSER = SimpleOptionParser
 
-    def __init__(self, go_dict=None, descr=None, short_groupdescr=None, long_groupdescr=None, config_files=None):
+    def __init__(self, go_dict=None, short_groupdescr=None, long_groupdescr=None, config_files=None):
```
Comment (JensTimmerman) [inline lib/vsc/utils/generaloption.py]: removing this argument will definitely break a lot of downstream binaries using this.
better `deprecate` it first, like, printing out that it's not used?

Comment (stdweird) [inline lib/vsc/utils/generaloption.py]: i actually grepped our code, and things look ok. but i'll reintroduce it.

File: lib/vsc/utils/affinity.py (hunk 3 above)
Comment (stdweird) [inline lib/vsc/utils/affinity.py]: will need to be fixed for py3

File: lib/vsc/utils/affinity.py (hunk 4 above)
Comment (stdweird) [inline lib/vsc/utils/affinity.py]: the remakr above it says 'tobin not used anymore'
i'll grep code and then remove it

File: lib/vsc/utils/affinity.py (hunk 3 above)
Comment (JensTimmerman) [inline lib/vsc/utils/affinity.py]: sure, but py2to3 does this. range creates a list, xrange creates a generator

**Comment by JensTimmerman** [comment]
jenkins test this please

**Comment by JensTimmerman** [comment]
jenkins test this please

File: test/run.py [hunk 6]
```diff
@@ -67,14 +77,79 @@ def test_simple_asyncloop(self):
...
+                try:
+                    os.kill(pid, 0) # test first
+                    os.kill(pid, 9)
+                except OSError:
+                    pass
+
+        do_test(False)
+        # TODO: find a way to change the pid group before starting this test
+        #       now it kills the test process too ;)
+        #       It's ok not to test, as it is not the default, and it's not easy to change it
```
Comment (JensTimmerman) [inline test/run.py]: you can for to get a new pid?

**Comment by stdweird** [comment]
test this please

**Comment by stdweird** [comment]
test this please

**Comment by stdweird** [comment]
test this please

**Comment by stdweird** [comment]
test this please

File: lib/vsc/utils/generaloption.py [hunk 7]
```diff
@@ -67,7 +67,7 @@ def set_columns(cols=None):
         if os.path.exists(stty):
             try:
                 cols = int(os.popen('%s size 2>/dev/null' % stty).read().strip().split(' ')[1])
-            except:
+            except (ValueError, OSError, AttributeError):
```
Comment (boegel) [inline lib/vsc/utils/generaloption.py]: @stdweird we missed `IndexError` here, cfr. #237 and #238

**Pull request merged**